### ISSUE TITLE
Commitments checked only in ProcessProposal

### ISSUE THREAD

**Issue by mpoke** [informalsystems/audit-celestia#30]
The [check](https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/app/process_proposal.go#L70) that PFD commitments are subtree roots of the data root is done only in `ProcessProposal`.  [ValidateBasic](https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/x/payment/types/payfordata.go#L39), which is executed on `DeliverTx`, does not validate the commitment (besides checking that it's not empty). IIUC, this check is essential for validating PFDs, as otherwise, it is not possible to check message inclusion. 

At the moment, the spec says only validators run `ProcessProposal`. In the implementation though, all nodes run `ProcessProposal`. If this changes in the future, then non-validator nodes will have to rely on the assumption that a majority of consensus is honest.

**Comment by rootulp** [comment]
[MsgPayForData](https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/x/payment/types/tx.pb.go#L214) doesn't contain the data so it's not possible to re-calculate the share commitments in MsgPayForData's `ValidateBasic()`.

WireMsgPayForData does have access to the data so it is possible to re-calculate and verify the share commitments in its `ValidateBasic()` [here](https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/x/payment/types/wirepayfordata.go#L105).

**Comment by mpoke** [comment]
> [MsgPayForData](https://github.com/celestiaorg/celestia-app/blob/e088d61fcb6579b4bc797deefd2ceff7601aa079/x/payment/types/tx.pb.go#L214) doesn't contain the data so it's not possible to re-calculate the share commitments in MsgPayForData's `ValidateBasic()`.

Isn't `MessageShareCommitment` the commitment?

**Comment by rootulp** [comment]
> Isn't `MessageShareCommitment` the commitment?

Yep. `MessageShareCommitment` is the commitment. But without the original data, it isn't possible to re-calculate the subtree roots that were used to create the commitment so it doesn't seem possible to perform this check

> that PFD commitments are subtree roots of the data root

as a stateless check in MsgPayForData's `ValidateBasic()`. I suppose it's possible to perform a stateful check in MsgPayForData's `ValidateBasic()` if provided information on the block to verify that a MsgPayForData was included in but I don't think that's what this issue is proposing. Is it?

**Comment by evan-forbes** [comment]
Yes, this is something that I was curious about in the slack channel. Nodes that sync the chain are not currently checking for message inclusion. It is currently possible for a malicious two thirds to produce a block without a message, and full nodes that sync won't notice. If they have a two thirds majority, then they can also simply withhold data, which would force all the other non-validator nodes to block sync and therefore miss this check. 

fortunately, we're planning on creating message inclusion fraud proof nodes that do not run tendermint. We still might want to fix this tho. I don't think it would be that difficult of a change. We simply pass the block data and header to ProcessProposal when syncing
